### PR TITLE
CIRC-1980: Error message - after Service point changed from "yes" to "no" as Pickup location

### PR DESCRIPTION
**Pull request by Maksat-Galymzhan**
## Purpose
[CIRC-1980](https://folio-org.atlassian.net/browse/CIRC-1980)

## Approach
Fix intervalId absence in hold shelf expiry period

### CODE REVIEW

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 1]
```diff
@@ -1662,6 +1665,60 @@ void linkItemToHoldTLRWithHoldShelfWhenCheckedInItemThenFulfilledWithSuccess(){
...
+    ServicePointBuilder changedServicePoint = new ServicePointBuilder(
+            servicePointsFixture.cd1().getId(),
+            servicePointName,
+            servicePointCode,
+            discoveryDisplayName,
+            description,
+            shelvingLagTime,
+            Boolean.FALSE,
+            null,
+            ExpirationDateManagement.KEEP_THE_CURRENT_DUE_DATE.name()
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckInByBarcodeTests.java:1692]: formatting

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 2]
```diff
@@ -1662,6 +1665,60 @@ void linkItemToHoldTLRWithHoldShelfWhenCheckedInItemThenFulfilledWithSuccess(){
...
+    );
+
+//    Update existing service point
```
Comment (alexanderkurash) [inline src/test/java/api/loans/CheckInByBarcodeTests.java:1695]: formatting (compare to comments formatting below)

File: src/test/java/api/support/fixtures/RecordCreator.java [hunk 3]
```diff
@@ -49,6 +49,10 @@ IndividualResource createIfAbsent(JsonObject record) {
     return createIfAbsent(identityMapKey.apply(record), record);
   }
 
+  void update(String servicePointCode, JsonObject updateRecord) {
+    identityMap.put(servicePointCode, client.create(updateRecord));
+  }
```
Comment (alexanderkurash) [inline src/test/java/api/support/fixtures/RecordCreator.java:54]: This class looks to be very generic, it shouldn't know anything about service point codes

File: src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java [hunk 4]
```diff
@@ -184,8 +185,15 @@ private RequestQueue setHoldShelfExpirationDateWithExpirationDateManagement(
 
     ExpirationDateManagement expirationDateManagement = calculatedRequest.getPickupServicePoint()
       .getHoldShelfClosedLibraryDateManagement();
-    String intervalId = calculatedRequest.getPickupServicePoint().getHoldShelfExpiryPeriod()
-      .getIntervalId().toUpperCase();
+
+    String intervalId = "0";
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java:189]: Can you use another value instead of this magic number?

Comment (Maksat-Galymzhan) [reply src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java]: updated

File: src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java [hunk 5]
```diff
@@ -184,8 +186,17 @@ private RequestQueue setHoldShelfExpirationDateWithExpirationDateManagement(
...
     ExpirationDateManagement expirationDateManagement = calculatedRequest.getPickupServicePoint()
       .getHoldShelfClosedLibraryDateManagement();
-    String intervalId = calculatedRequest.getPickupServicePoint().getHoldShelfExpiryPeriod()
-      .getIntervalId().toUpperCase();
+
+    String intervalId;
+    TimePeriod holdShelfExpiryPeriod = calculatedRequest.getPickupServicePoint()
+      .getHoldShelfExpiryPeriod();
+    if (holdShelfExpiryPeriod != null && isNotBlank(holdShelfExpiryPeriod.getIntervalId())) {
+      intervalId = holdShelfExpiryPeriod.getIntervalId().toUpperCase();
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java:194]: Please add logging here as well

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 6]
```diff
@@ -1662,6 +1665,51 @@ void linkItemToHoldTLRWithHoldShelfWhenCheckedInItemThenFulfilledWithSuccess(){
...
   }
 
+  @Test
+  void checkInItemWhenServiceHasChangedToNoPickupLocation(){
+    reconfigureTlrFeature(TlrFeatureStatus.NOT_CONFIGURED);
+    configurationsFixture.enableTlrFeature();
+    UUID instanceId = instancesFixture.basedUponDunkirk().getId();
+    IndividualResource defaultWithHoldings = holdingsFixture.defaultWithHoldings(instanceId);
+    IndividualResource checkedOutItem = itemsClient.create(buildCheckedOutItemWithHoldingRecordsId(defaultWithHoldings.getId()));
+    IndividualResource holdRequestBeforeFulfilled = requestsClient.create(buildHoldTLRWithHoldShelffulfillmentPreference(instanceId));
```
Comment (roman-barannyk) [inline src/test/java/api/loans/CheckInByBarcodeTests.java:1675]:
```suggestion
    var instanceId = instancesFixture.basedUponDunkirk().getId();
    var defaultWithHoldings = holdingsFixture.defaultWithHoldings(instanceId);
    var checkedOutItem = itemsClient.create(buildCheckedOutItemWithHoldingRecordsId(defaultWithHoldings.getId()));
    var holdRequestBeforeFulfilled = requestsClient.create(buildHoldTLRWithHoldShelffulfillmentPreference(instanceId));
```

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 7]
```diff
@@ -1662,6 +1665,51 @@ void linkItemToHoldTLRWithHoldShelfWhenCheckedInItemThenFulfilledWithSuccess(){
...
+    assertThat(representation.getString("holdingsRecordId"), is(defaultWithHoldings.getId()));
+
+    validateTLRequestByFields(representation, HOLD_SHELF, instanceId, OPEN_AWAITING_PICKUP);
+
+    IndividualResource itemAfter = itemsClient.get(checkedOutItem.getId());
+    JsonObject itemRepresentation = itemAfter.getJson();
+    assertThat(itemRepresentation.getJsonObject("status").getString("name"), is("Awaiting pickup"));
+
+    JsonObject servicePointRepresentation = representation.getJsonObject("pickupServicePoint");
+    Assertions.assertFalse(Boolean.parseBoolean(servicePointRepresentation.getString("pickupLocation")));
```
Comment (roman-barannyk) [inline src/test/java/api/loans/CheckInByBarcodeTests.java:1710]: would suggest using the same import for assertions as seen throughout this test

File: src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java [hunk 8]
```diff
@@ -184,8 +186,17 @@ private RequestQueue setHoldShelfExpirationDateWithExpirationDateManagement(
...
+    } else {
+      log.info("setHoldShelfExpirationDateWithExpirationDateManagement:: hold shelf expiry period interval is empty");
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java:196]: please use limit of 100 symbols according to folio check style

File: src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java (hunk 5 above)
Comment (Maksat-Galymzhan) [reply src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java:195]: done

File: src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Maksat-Galymzhan) [reply src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java]: done

File: src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java [hunk 9]
```diff
@@ -184,8 +186,18 @@ private RequestQueue setHoldShelfExpirationDateWithExpirationDateManagement(
...
     ExpirationDateManagement expirationDateManagement = calculatedRequest.getPickupServicePoint()
       .getHoldShelfClosedLibraryDateManagement();
-    String intervalId = calculatedRequest.getPickupServicePoint().getHoldShelfExpiryPeriod()
-      .getIntervalId().toUpperCase();
+
+    String intervalId;
+    TimePeriod holdShelfExpiryPeriod = calculatedRequest.getPickupServicePoint()
+      .getHoldShelfExpiryPeriod();
+    if (holdShelfExpiryPeriod != null && isNotBlank(holdShelfExpiryPeriod.getIntervalId())) {
+      log.info("setHoldShelfExpirationDateWithExpirationDateManagement:: interval is {}", holdShelfExpiryPeriod.getIntervalId());
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java:194]:
```suggestion
      log.info("setHoldShelfExpirationDateWithExpirationDateManagement:: interval is {}", 
        holdShelfExpiryPeriod.getIntervalId());
```

**Review by roman-barannyk — APPROVED** [review]

File: src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java [hunk 10]
```diff
@@ -184,8 +186,19 @@ private RequestQueue setHoldShelfExpirationDateWithExpirationDateManagement(
 
     ExpirationDateManagement expirationDateManagement = calculatedRequest.getPickupServicePoint()
       .getHoldShelfClosedLibraryDateManagement();
-    String intervalId = calculatedRequest.getPickupServicePoint().getHoldShelfExpiryPeriod()
-      .getIntervalId().toUpperCase();
+
+    String intervalId;
+    TimePeriod holdShelfExpiryPeriod = calculatedRequest.getPickupServicePoint()
+      .getHoldShelfExpiryPeriod();
+    if (holdShelfExpiryPeriod != null && isNotBlank(holdShelfExpiryPeriod.getIntervalId())) {
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java:193]: You are calling `holdShelfExpiryPeriod.getIntervalId()` three times here. How about an `Optional`?

```
    String intervalId = Optional.of(calculatedRequest)
      .map(Request::getPickupServicePoint)
      .map(ServicePoint::getHoldShelfExpiryPeriod)
      .map(TimePeriod::getIntervalId)
      .map(String::toUpperCase)
      .orElse(NOT_DEFINED_INTERVAL);

    log.info("setHoldShelfExpirationDateWithExpirationDateManagement:: interval: {}", intervalId);
```

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 11]
```diff
@@ -1662,6 +1664,55 @@ void linkItemToHoldTLRWithHoldShelfWhenCheckedInItemThenFulfilledWithSuccess(){
     assertThat(itemRepresentation.getJsonObject("status").getString("name"), is("Awaiting pickup"));
   }
 
+  @Test
+  void checkInItemWhenServiceHasChangedToNoPickupLocation(){
+    reconfigureTlrFeature(TlrFeatureStatus.NOT_CONFIGURED);
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/CheckInByBarcodeTests.java:1669]: No need to do this, `configurationsFixture.enableTlrFeature()` will take care of wiping existing configuration

File: src/test/java/api/loans/CheckInByBarcodeTests.java [hunk 12]
```diff
@@ -1662,6 +1664,55 @@ void linkItemToHoldTLRWithHoldShelfWhenCheckedInItemThenFulfilledWithSuccess(){
     assertThat(itemRepresentation.getJsonObject("status").getString("name"), is("Awaiting pickup"));
   }
 
+  @Test
+  void checkInItemWhenServiceHasChangedToNoPickupLocation(){
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/loans/CheckInByBarcodeTests.java:1668]:
```suggestion
  void checkInItemWhenServicePointHasChangedToNoPickupLocation(){
```

File: src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java (hunk 10 above)
Comment (Maksat-Galymzhan) [reply src/main/java/org/folio/circulation/domain/UpdateRequestQueue.java]: done

File: src/test/java/api/loans/CheckInByBarcodeTests.java (hunk 11 above)
Comment (Maksat-Galymzhan) [reply src/test/java/api/loans/CheckInByBarcodeTests.java]: done

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**